File: src/core/geometry/tetrahedron/python/TetrahedronCreateTetId.py
def createTetIds(verts, tree, minQuality):

    tetIds = []
    neighbors = []
    tetMarks = []
    tetMark = 0
    firstFreeTet = -1

    planesN = []
    planesD = []

    firstBig = len(verts) - 4

    # first big tet
    tetIds.append(firstBig)
    tetIds.append(firstBig + 1)
    tetIds.append(firstBig + 2)
    tetIds.append(firstBig + 3)
    tetMarks.append(0)

    for i in range(4):
        neighbors.append(-1)
        p0 = verts[firstBig + tetFaces[i][0]]
        p1 = verts[firstBig + tetFaces[i][1]]
        p2 = verts[firstBig + tetFaces[i][2]]
        n = (p1 - p0).cross(p2 - p0)
        n.normalize()
        planesN.append(n)
        planesD.append(p0.dot(n))

    center = mathutils.Vector((0.0, 0.0, 0.0))

    print(" ------------- tetrahedralization ------------------- ")
    
    for i in range(0, firstBig):
        p = verts[i]

        if i % 100 == 0:
            print("inserting vert", i + 1, "of", firstBig)

        # find non-deleted tet

        tetNr = 0
        while tetIds[4 * tetNr] < 0:
            tetNr = tetNr + 1
            
        # find containing tet

        tetMark = tetMark + 1
        found = False

        while not found:
            if tetNr < 0 or tetMarks[tetNr] == tetMark:
                break
            tetMarks[tetNr] = tetMark

            id0 = tetIds[4 * tetNr]
            id1 = tetIds[4 * tetNr + 1]
            id2 = tetIds[4 * tetNr + 2]
            id3 = tetIds[4 * tetNr + 3]

            center = (verts[id0] + verts[id1] + verts[id2] + verts[id3]) * 0.25

            minT = float('inf')
            minFaceNr = -1

            for j in range(0, 4):
                n = planesN[4 * tetNr + j]
                d = planesD[4 * tetNr + j]

                hp = n.dot(p) - d
                hc = n.dot(center) - d

                t = hp - hc
                if t == 0:
                    continue

                # time when c -> p hits the face
                t = -hc / t

                if t >= 0.0 and t < minT:
                    minT = t
                    minFaceNr = j
            
            if minT >= 1.0:
                found = True
            else:
                tetNr = neighbors[4 * tetNr + minFaceNr]
        
        if not found:
            print("*********** failed to insert vertex")
            continue
        
        # find violating tets

        tetMark = tetMark + 1

        violatingTets = []
        stack = [tetNr]

        while len(stack) != 0:
            tetNr = stack.pop()
            if tetMarks[tetNr] == tetMark:
                continue
            tetMarks[tetNr] = tetMark
            violatingTets.append(tetNr)

            for j in range(4):
                n = neighbors[4 * tetNr + j]
                if n < 0 or tetMarks[n] == tetMark:
                    continue
                
                # Delaunay condition test

                id0 = tetIds[4 * n]
                id1 = tetIds[4 * n + 1]
                id2 = tetIds[4 * n + 2]
                id3 = tetIds[4 * n + 3]

                c = getCircumCenter(verts[id0], verts[id1], verts[id2], verts[id3])

                r = (verts[id0] - c).magnitude
                if (p - c).magnitude < r:
                    stack.append(n)

        # remove old tets, create new ondes

        edges = []

        for j in range(len(violatingTets)):
            tetNr = violatingTets[j]

            # copy info before we delete it
            ids = [0] * 4
            ns = [0] * 4
            for k in range(4):
                ids[k] = tetIds[4 * tetNr + k]
                ns[k] = neighbors[4 * tetNr + k]

            # delete the tet
            tetIds[4 * tetNr] = -1
            tetIds[4 * tetNr + 1] = firstFreeTet
            firstFreeTet = tetNr

            # visit neighbors

            for k in range(4):
                n = ns[k]
                if n >= 0 and tetMarks[n] == tetMark:
                    continue

                # no neighbor or neighbor is not-violating -> we are facing the border

                # create new tet

                newTetNr = firstFreeTet
                
                if newTetNr >= 0:
                    firstFreeTet = tetIds[4 * firstFreeTet + 1]
                else:
                    newTetNr = int(len(tetIds) / 4)
                    tetMarks.append(0)
                    for l in range(4):
                        tetIds.append(-1)
                        neighbors.append(-1)
                        planesN.append(mathutils.Vector((0.0, 0.0, 0.0)))
                        planesD.append(0.0)

                id0 = ids[tetFaces[k][2]]
                id1 = ids[tetFaces[k][1]]
                id2 = ids[tetFaces[k][0]]

                tetIds[4 * newTetNr] = id0
                tetIds[4 * newTetNr + 1] = id1
                tetIds[4 * newTetNr + 2] = id2
                tetIds[4 * newTetNr + 3] = i

                neighbors[4 * newTetNr] = n

                if n >= 0:
                    for l in range(4):
                        if neighbors[4 * n + l] == tetNr:
                            neighbors[4 * n + l] = newTetNr
                        
                # will set the neighbors among the new tets later

                neighbors[4 * newTetNr + 1] = -1
                neighbors[4 * newTetNr + 2] = -1
                neighbors[4 * newTetNr + 3] = -1

                for l in range(4):
                    p0 = verts[tetIds[4 * newTetNr + tetFaces[l][0]]]
                    p1 = verts[tetIds[4 * newTetNr + tetFaces[l][1]]]
                    p2 = verts[tetIds[4 * newTetNr + tetFaces[l][2]]]
                    newN = (p1 - p0).cross(p2 - p0)
                    newN.normalize()
                    planesN[4 * newTetNr + l] = newN
                    planesD[4 * newTetNr + l] = newN.dot(p0)

                if id0 < id1:
                    edges.append((id0, id1, newTetNr, 1))
                else:
                    edges.append((id1, id0, newTetNr, 1))

                if id1 < id2:
                    edges.append((id1, id2, newTetNr, 2))
                else:
                    edges.append((id2, id1, newTetNr, 2))

                if id2 < id0:
                    edges.append((id2, id0, newTetNr, 3))
                else:
                    edges.append((id0, id2, newTetNr, 3))

            # next neighbor
        # next violating tet

        # fix neighbors

        sortedEdges = sorted(edges, key = cmp_to_key(compareEdges))

        nr = 0
        numEdges = len(sortedEdges)

        while nr < numEdges:
            e0 = sortedEdges[nr]
            nr = nr + 1

            if nr < numEdges and equalEdges(sortedEdges[nr], e0):
                e1 = sortedEdges[nr]

                id0 = tetIds[4 * e0[2]]
                id1 = tetIds[4 * e0[2] + 1]
                id2 = tetIds[4 * e0[2] + 2]
                id3 = tetIds[4 * e0[2] + 3]

                jd0 = tetIds[4 * e1[2]]
                jd1 = tetIds[4 * e1[2] + 1]
                jd2 = tetIds[4 * e1[2] + 2]
                jd3 = tetIds[4 * e1[2] + 3]

                neighbors[4 * e0[2] + e0[3]] = e1[2]
                neighbors[4 * e1[2] + e1[3]] = e0[2]
                nr = nr + 1

    # next point

    # remove outer, deleted and outside tets

    numTets = int(len(tetIds) / 4)
    num = 0
    numBad = 0

    for i in range(numTets):
        id0 = tetIds[4 * i]
        id1 = tetIds[4 * i + 1]
        id2 = tetIds[4 * i + 2]
        id3 = tetIds[4 * i + 3]

        if id0 < 0 or id0 >= firstBig or id1 >= firstBig or id2 >= firstBig or id3 >= firstBig:
            continue

        p0 = verts[id0]
        p1 = verts[id1]
        p2 = verts[id2]
        p3 = verts[id3]
        
        quality = tetQuality(p0, p1, p2, p3)
        if quality < minQuality:
            numBad = numBad + 1
            continue

        center = (p0 + p1 + p2 + p3) * 0.25
        if not isInside(tree, center):
            continue

        tetIds[num] = id0
        num = num + 1
        tetIds[num] = id1
        num = num + 1
        tetIds[num] = id2
        num = num + 1
        tetIds[num] = id3
        num = num + 1

    del tetIds[num:]

    print(numBad, "bad tets deleted")
    print(int(len(tetIds) / 4),"tets created")

    return tetIds

File: src/core/geometry/tetrahedron/TetrahedronCreateTetId.ts
// import {Matrix4, Vector3} from 'three';
// import {TetCreationStage, TetEdge} from './TetrahedronConstant';
// import {MeshWithBVHGeometry} from '../bvh/ThreeMeshBVHHelper';

// import {_findContainingTet} from './utils/TetIdFindContainingTet';
// import {_visitNeighbours} from './utils/TetIdVisitNeighbours';
// import {_fixNeighbours} from './utils/TetIdFixNeighbours';
// import {_removeOutsideTets} from './utils/TetIdRemoveOutsideTets';
// import {_findViolatingTets} from './utils/TetIdFindViolatingTets';
// import {TET_FACE_POINT_INDICES} from '../tet/TetCommon';

// const _p = new Vector3();
// const _p0 = new Vector3();
// const _p1 = new Vector3();
// const _p2 = new Vector3();
// const _center = new Vector3();

// interface CreateTetIdsOptions {
// 	invMat: Matrix4;
// 	mesh: MeshWithBVHGeometry;
// 	verts: Vector3[];
// 	minQuality: number;
// 	//
// 	stage: TetCreationStage;
// 	subStage: number;
// 	removeOutsideTets: boolean;
// }
// interface CreateTetIdsResult {
// 	tetIds: number[];
// }

// export function createTetIds(options: CreateTetIdsOptions): CreateTetIdsResult {
// 	const {invMat, mesh, verts, minQuality, stage, subStage, removeOutsideTets} = options;
// 	const tetIds: number[] = [];
// 	const neighbors: number[] = [];
// 	const tetMarks = [];

// 	let tetMark: number = 0;
// 	let firstFreeTet: number = -1;

// 	const planesN: Vector3[] = [];
// 	const planesD: number[] = [];

// 	const firstBig = verts.length - 4;

// 	// first big tet
// 	tetIds.push(firstBig);
// 	tetIds.push(firstBig + 1);
// 	tetIds.push(firstBig + 2);
// 	tetIds.push(firstBig + 3);
// 	if (stage == TetCreationStage.TETS && subStage == 0) {
// 		return {tetIds};
// 	}

// 	tetMarks.push(0);

// 	// for i in range(4):
// 	//     neighbors.append(-1)
// 	//     p0 = verts[firstBig + tetFaces[i][0]]
// 	//     p1 = verts[firstBig + tetFaces[i][1]]
// 	//     p2 = verts[firstBig + tetFaces[i][2]]
// 	//     n = (p1 - p0).cross(p2 - p0)
// 	//     n.normalize()
// 	//     planesN.append(n)
// 	//     planesD.append(p0.dot(n))
// 	for (let i = 0; i < 4; i++) {
// 		neighbors.push(-1);
// 		_p0.copy(verts[firstBig + TET_FACE_POINT_INDICES[i][0]]);
// 		_p1.copy(verts[firstBig + TET_FACE_POINT_INDICES[i][1]]);
// 		_p2.copy(verts[firstBig + TET_FACE_POINT_INDICES[i][2]]);
// 		_p1.sub(_p0);
// 		_p2.sub(_p0);
// 		const n = new Vector3().copy(_p1).cross(_p2);
// 		n.normalize();
// 		planesN.push(n);
// 		planesD.push(_p0.dot(n));
// 	}

// 	_center.set(0.0, 0.0, 0.0);

// 	console.warn(' ------------- tetrahedralization ------------------- ');

// 	for (let i = 0; i < firstBig; i++) {
// 		if (i > subStage) {
// 			break;
// 		}
// 		_p.copy(verts[i]);
// 		// if (i % 100 == 0) {
// 		console.log('inserting vert', i, 'of', firstBig);
// 		// }
// 		// find non-deleted tet
// 		let tetNr: number = 0;
// 		while (tetIds[4 * tetNr] < 0) {
// 			tetNr++;
// 		}

// 		// find containing tet
// 		tetMark++;
// 		const result = _findContainingTet({
// 			vertPos: _p,
// 			tetIds,
// 			tetMarks,
// 			tetMark,
// 			verts,
// 			tetNr,
// 			neighbors,
// 			planesN,
// 			planesD,
// 		});
// 		tetNr = result.tetNr;
// 		if (!result.found) {
// 			console.warn(`*********** failed to insert vertex '${i}'`);
// 			continue;
// 		}

// 		// find violating tets

// 		tetMark++;
// 		// const violatingTets: number[] = [];
// 		// let stack = [tetNr];

// 		// while (stack.length != 0) {
// 		// 	tetNr = stack.pop()!;
// 		// 	if (tetMarks[tetNr] == tetMark) {
// 		// 		continue;
// 		// 	}
// 		// 	tetMarks[tetNr] = tetMark;
// 		// 	violatingTets.push(tetNr);

// 		// 	for (let j = 0; j < 4; j++) {
// 		// 		const n = neighbors[4 * tetNr + j];
// 		// 		if (n < 0 || tetMarks[n] == tetMark) {
// 		// 			continue;
// 		// 		}
// 		// 		// Delaunay condition test
// 		// 		const id0 = tetIds[4 * n];
// 		// 		const id1 = tetIds[4 * n + 1];
// 		// 		const id2 = tetIds[4 * n + 2];
// 		// 		const id3 = tetIds[4 * n + 3];

// 		// 		getCircumCenter(verts[id0], verts[id1], verts[id2], verts[id3], _c);

// 		// 		const r = _v3.copy(verts[id0]).sub(_c).length();
// 		// 		if (_v3.copy(_p).sub(_c).length() < r) {
// 		// 			stack.push(n);
// 		// 		}
// 		// 	}
// 		// }
// 		const violatingTestsResult = _findViolatingTets({
// 			vertPos: _p,
// 			tetIds,
// 			tetMarks,
// 			tetMark,
// 			verts,
// 			tetNr,
// 			neighbors,
// 		});
// 		const violatingTets = violatingTestsResult.violatingTets;
// 		tetNr = violatingTestsResult.tetNr;

// 		// remove old tets, create new ones

// 		const edges: TetEdge[] = [];
// 		const violatingTetsCount = violatingTets.length;
// 		for (let j = 0; j < violatingTetsCount; j++) {
// 			tetNr = violatingTets[j];
// 			const ids = [0, 0, 0, 0];
// 			const ns = [0, 0, 0, 0];
// 			for (let k = 0; k < 4; k++) {
// 				ids[k] = tetIds[4 * tetNr + k];
// 				ns[k] = neighbors[4 * tetNr + k];
// 			}
// 			// delete the tet
// 			tetIds[4 * tetNr] = -1;
// 			tetIds[4 * tetNr + 1] = firstFreeTet;
// 			firstFreeTet = tetNr;

// 			// visit neighbors
// 			const result = _visitNeighbours({
// 				tetMarks,
// 				tetMark,
// 				tetIds,
// 				tetNr,
// 				edges,
// 				neighbors,
// 				ns,
// 				firstFreeTet,
// 				verts,
// 				planesN,
// 				planesD,
// 				ids,
// 				i,
// 			});
// 			firstFreeTet = result.firstFreeTet;

// 			// next neighbor

// 			// next violating tet

// 			// fix neighbors

// 			_fixNeighbours({edges, neighbors});
// 		}

// 		// next point
// 	}
// 	// remove outer, deleted and outside tets

// 	const newTetIds = removeOutsideTets
// 		? _removeOutsideTets({
// 				invMat,
// 				mesh,
// 				tetIds,
// 				firstBig,
// 				verts,
// 				minQuality,
// 		  })
// 		: tetIds;

// 	console.warn(Math.floor(newTetIds.length / 4), 'tets created');

// 	return {tetIds: newTetIds};
// }
